galley/pkg/runtime/processor.go

```diff
@@ -77,90 +71,63 @@ func newProcessor(
 		state:           state,
 		source:          src,
 		postProcessHook: postProcessHook,
-		done:            make(chan struct{}),
-		stopped:         make(chan struct{}),
+		worker:          util.NewWorker("runtime processor", scope),
 		lastEventTime:   now,
 	}
 }
 
 // Start the processor. This will cause processor to listen to incoming events from the provider
 // and publish component configuration via the Distributor.
 func (p *Processor) Start() error {
-	scope.Info("Starting processor...")
-
-	if p.events != nil {
-		scope.Warn("Processor has already started")
-		return errors.New("already started")
-	}
-
-	events := make(chan resource.Event, 1024)
-	err := p.source.Start(func(e resource.Event) {
-		events <- e
+	return p.worker.Start(func(ctx context.Context) {
+		scope.Info("Starting processor...")
+		eventCh := make(chan resource.Event, 1024)
+		defer func() {
+			scope.Debugf("Process.process: Exiting worker thread")
+			close(eventCh)
+			p.state.close()
+		}()
+
+		err := p.source.Start(func(e resource.Event) {
+			eventCh <- e
+		})
+		if err != nil {
```

Comment:
I don't think this is right. I thought I mentioned this in our hangouts chat, but the source.Start returning an error is a legitimate problem that should be bubbled up correctly. double-start/wrong argument are things we can panic on.